cpu/stm32l1/stmclk.c

```diff
@@ -0,0 +1,242 @@
+#if (CLOCK_CORECLOCK > 8000000U)
+    /* Enable 64-bit access, prefetch and 1 wait state */
+    /* (at F > 8MHz/16MHz WS must be 1) */
+    FLASH->ACR |= FLASH_ACR_ACC64;
+    FLASH->ACR |= CLOCK_FLASH_LATENCY;
+    FLASH->ACR |= FLASH_ACR_PRFTEN;
+    /* Wait for flash to become ready */
+    while (!(FLASH->SR & FLASH_SR_READY)) {}
+#else
+    /* Set 0 wait state, 32-bit access and no prefetch */
+    /* LATENCY and PRFTEN can be changed with 64-bit access enabled only */
+    FLASH->ACR |= FLASH_ACR_ACC64;
+    FLASH->ACR &= ~FLASH_ACR_LATENCY;
+    FLASH->ACR &= ~FLASH_ACR_PRFTEN;
+    FLASH->ACR &= ~FLASH_ACR_ACC64;
+    /* Wait for flash to become ready */
+    while (!(FLASH->SR & FLASH_SR_READY)) {}
+#endif
+
+    /* Power domain enable */
+    periph_clk_en(APB1, RCC_APB1ENR_PWREN);
+    /* Select the Voltage Range */
+    tmpreg = PWR->CR;
+    tmpreg &= ~PWR_CR_VOS;
+    tmpreg |= CORE_VOLTAGE;
+    PWR->CR = tmpreg;
+    /* Wait until the Voltage Regulator is ready */
+    while((PWR->CSR & PWR_CSR_VOSF) != 0) {}
+
+    /* set AHB, APB1 and APB2 clock dividers */
+    tmpreg = RCC->CFGR;
+    tmpreg &= ~RCC_CFGR_HPRE;
+    tmpreg |= (uint32_t)CLOCK_AHB_DIV;
+    tmpreg &= ~RCC_CFGR_PPRE1;
+    tmpreg |= (uint32_t)CLOCK_APB1_DIV;
+    tmpreg &= ~RCC_CFGR_PPRE2;
+    tmpreg |= (uint32_t)CLOCK_APB2_DIV;
+    RCC->CFGR = tmpreg;
+
+#if CLOCK_USE_PLL
+    /*  PLL configuration: PLLCLK = CLOCK_SOURCE / PLL_DIV * PLL_MUL */
+#if defined(CLOCK_PLL_MULTI)
+    if (clock_source_rdy == RCC_CR_HSERDY) {
+        RCC->CFGR |= (uint32_t)(RCC_CFGR_PLLSRC_HSE | CLOCK_PLL_DIV_HSE | CLOCK_PLL_MUL_HSE);
+    } else {
+        RCC->CFGR |= (uint32_t)(RCC_CFGR_PLLSRC_HSI | CLOCK_PLL_DIV_HSI | CLOCK_PLL_MUL_HSI);
+    }
+#else
+    RCC->CFGR |= (uint32_t)(CLOCK_PLL_SOURCE | CLOCK_PLL_DIV | CLOCK_PLL_MUL);
+#endif
+    /* Enable PLL */
+    RCC->CR |= RCC_CR_PLLON;
+    /* Wait till PLL is ready */
+    while ((RCC->CR & RCC_CR_PLLRDY) == 0) {}
+#endif
+
+    /* Select system clock source and turn of other clock sources*/
+    tmpreg = RCC->CFGR;
+    tmpreg &= ~RCC_CFGR_SW;
+
+    uint32_t clock_cfgr_sw;
+    uint32_t clock_cfgr_sw_rdy;
+    uint32_t clock_disable_clocks;
+
+    if (clock_source_rdy == RCC_CR_HSERDY) {
+        clock_cfgr_sw = RCC_CFGR_SW_HSE;
+        clock_cfgr_sw_rdy = RCC_CFGR_SWS_HSE;
+        clock_disable_clocks = RCC_CR_HSION | RCC_CR_MSION;
+    } else {
+        clock_cfgr_sw = RCC_CFGR_SW_HSI;
+        clock_cfgr_sw_rdy = RCC_CFGR_SWS_HSI;
+        clock_disable_clocks = RCC_CR_HSEON | RCC_CR_MSION;
+    }
+    if (CLOCK_USE_PLL) {
+        clock_cfgr_sw = RCC_CFGR_SW_PLL;
+        clock_cfgr_sw_rdy = RCC_CFGR_SWS_PLL;
+    }
+
+    tmpreg |= (uint32_t)clock_cfgr_sw;
+    RCC->CFGR = tmpreg;
+
+    /* Wait for sysyem clock to be ready before desabling other clocks*/
+    while ((RCC->CFGR & (uint32_t)RCC_CFGR_SWS) != clock_cfgr_sw_rdy) {}
+    RCC->CR &= ~(clock_disable_clocks);
```

Comment:
And here `stmclk_disable_hsi();` instead.

Reply:
needed for #8403 to switch to low-power sleep mode

Reply:
it's not always HSI, can be HSI+MSI or HSE+MSI.